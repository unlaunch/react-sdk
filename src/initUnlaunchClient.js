import  {default as ULClient} from "unlaunch-js-client-lib";
/**
 * Internal function to initialize the `ULClient`.
 */
const initUnlaunchClient = async (
  apiKey,
  flagKeySet,
  identity = 'anonymous',
  attributes = null,
  options={}
) => {
  const unlaunchClient = ULClient.initialize(
    apiKey,
    flagKeySet,
    identity,
    null,
    options
  );
  return new Promise(resolve => {

    unlaunchClient.on('ready', function () {

      let rawFlags = {};

      if (flagKeySet) {
        for (const flag in flagKeySet) {
          rawFlags[flagKeySet[flag]] = unlaunchClient.variation(flagKeySet[flag]);
        }
      }
      const flags = rawFlags;
      resolve({ flags, unlaunchClient });
    });
  });
}

export default initUnlaunchClient;
