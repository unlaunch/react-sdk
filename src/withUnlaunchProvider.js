import * as React from 'react';
import UnlaunchProvider from './provider';

/**
 * `withUnlaunchProvider` is a function which accepts a config object which is used to
 * initialize `unlaunch-js-client-sdk`
 */
export function withUnlaunchProvider(config) {
  return function withUnlaunchProviderHoc(WrappedComponent) {

    return class extends React.Component{
      render() {
        return (
          <UnlaunchProvider {...config}>
            <WrappedComponent {...this.props} />
          </UnlaunchProvider>
        );
      }
    };
  };
}

export default withUnlaunchProvider;