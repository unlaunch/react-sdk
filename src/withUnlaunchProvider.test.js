jest.mock('./initUnlaunchClient', () => jest.fn());
jest.mock('./context', () => ({ Provider: 'Provider' }));

import * as React from 'react';
import { create } from 'react-test-renderer';
import initUnlaunchClient from './initUnlaunchClient';
import withUnlaunchProvider from './withUnlaunchProvider';
import UnlaunchProvider from './provider';

const apiKey = 'apiKey';
const App = () => <div>My App</div>;
const mockInitUnlaunchClient = initUnlaunchClient;
const mockFlags = { "test-flag-1": true, "test-flag-2": true };
const mockUnlaunchClient = {
  on: jest.fn((e, cb) => {
    cb();
  }),
};

describe('withUnlaunchProvider', () => {
  beforeEach(() => {
    mockInitUnlaunchClient.mockImplementation(() => ({
      flags: mockFlags,
      unlaunchClient: mockUnlaunchClient,
    }));
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  test('render app', () => {
    const UnlaunchApp = withUnlaunchProvider({ apiKey })(App);
    const component = create(<UnlaunchApp />);
    expect(component).toMatchSnapshot();
  });

  test('unlaunch client is initialized correctly', async () => {
    const identity = "user-123";
    const options = {};
    const UnlaunchApp = withUnlaunchProvider({ apiKey, flagKeys:undefined, options:{}, attributes:null,identity  })(App);
    const instance = create(<UnlaunchApp />).root.findByType(UnlaunchProvider).instance;

    await instance.componentDidMount();
    expect(mockInitUnlaunchClient).toHaveBeenCalledWith(apiKey, undefined, identity, null, {});
  });

  test('unlaunch client is initialized correctly with target flags', async () => {
    mockInitUnlaunchClient.mockImplementation(() => ({
      flags: { "test-flag-1": true, "test-flag-2": true },
      unlaunchClient: mockUnlaunchClient,
    }));
    const identity = "user-123";
    const options = {};
    const flagKeys = { 'test-flag-1': false, 'test-flag-2': false };
    const UnlaunchApp = withUnlaunchProvider({apiKey, flagKeys, options, attributes:null,identity })(App);
    const instance = create(<UnlaunchApp />).root.findByType(UnlaunchProvider).instance;
    instance.setState = jest.fn();

    await instance.componentDidMount();

    expect(mockInitUnlaunchClient).toHaveBeenCalledWith(apiKey, flagKeys, identity, null, {});
    expect(instance.setState).toHaveBeenCalledWith({
      flags: { "test-flag-1": true, "test-flag-2": true },
      unlaunchClient: mockUnlaunchClient,
    });
  });

  test('flags and unlaunchClient are saved in state on mount', async () => {
    const flagKeys = { 'test-flag-1': false, 'test-flag-2': false };
    const identity = "user-123";
    const UnlaunchApp = withUnlaunchProvider({ apiKey, flagKeys, options:{}, attributes:null,identity })(App);
    const instance = create(<UnlaunchApp />).root.findByType(UnlaunchProvider).instance;
    instance.setState = jest.fn();

    await instance.componentDidMount();
    expect(instance.setState).toHaveBeenCalledWith({ flags: mockFlags, unlaunchClient: mockUnlaunchClient });
  });
})