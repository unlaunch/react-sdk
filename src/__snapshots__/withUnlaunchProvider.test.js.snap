// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withUnlaunchProvider render app 1`] = `
<Provider
  value={
    Object {
      "flags": Object {},
      "unlaunchClient": undefined,
    }
  }
>
  <div>
    My App
  </div>
</Provider>
`;
